modules/data.remote/R/SMAP_SMP_prep.R

```diff
@@ -0,0 +1,96 @@
+#'
+#' @return A data frame containing SMAP smp and sd for each site and each time step.
+#' @export
+#'
+#' @examples
+#' @author Dongchen Zhang
+#' @importFrom magrittr %>%
+SMAP_SMP_prep <- function(site_info, start_date, end_date, time_points, 
+                      outdir, search_window = 30, export_csv = TRUE, update_csv = FALSE){
+  #note that, the SMAP_gee.csv file comes from Google Earth Engine (GEE) directly.
+  #Code for generating this file can be found through this link: 
+  #https://code.earthengine.google.com/ecbeb770e576d8ef72f72f5f12da3496
+  #Feel free to contact Dongchen Zhang (zhangdc@bu.edu) who wrote this code.
+  #The SMAP.csv file will be generated the first time when you use this code.
+  #for the next time, it will save you lot of time if you can provide the SMAP.csv directly.
+  
+  #check if SMAP.csv exists.
+  if(!file.exists(file.path(outdir, "SMAP.csv")) | as.logical(update_csv)){
+    if(!file.exists(file.path(outdir, "SMAP_gee.csv"))){
+      PEcAn.logger::logger.info("Please Provide SMAP dir that contains at least the SMAP_gee.csv file!")
+      return(0)
+    }else{
+      SMAP_gee <- utils::read.csv(file.path(outdir, "SMAP_gee.csv"))
+      
+      #prepare CSV from Current SMAP_gee file
+      SMAP_CSV <- matrix(NA, 0, 6) %>% `colnames<-`(c("date", "site_id", "lat", "lon", "smp", "sd"))
+      
+      for (i in 2 : dim(SMAP_gee)[1]) {
+        String <-  strsplit(gsub(",", "", gsub("\\[|\\]", "", SMAP_gee[i,2])), " ")[[1]]
+        date <- as.Date(strsplit(String[1], "_")[[1]][5], "%Y%m%d")
+        lon <- as.numeric(String[2])
+        lat <- as.numeric(String[3])
+        smp <- as.numeric(String[5]) * 100
+        sd <- 0.04 * 100 #From Daniel
+        
+        #Match current lon/lat with site_info
+        Distance <- swfscMisc::distance(lat1 = lat, lon1 = lon, lat2 = site_info$lat, lon2 = site_info$lon)
+        distloc <- match(min(Distance), Distance)
+        site_id <- site_info$site_id[distloc]
+        
+        SMAP_CSV <- rbind(SMAP_CSV, tibble::tibble(date, site_id, lat, lon, smp, sd))#in date, id, lat, lon, smp, sd
+      }
+      if(as.logical((export_csv))){
+        utils::write.csv(SMAP_CSV, file = file.path(outdir, "SMAP.csv"), row.names = F)
+      }
+    }
+  }else{
+    #TODO: When current SMAP.csv need to be updated
+    SMAP_CSV <- utils::read.csv(file.path(outdir, "SMAP.csv"))
+    Current_years <- sort(unique(lubridate::year(SMAP_CSV$date)))
+    Required_years <- lubridate::year(start_date):lubridate::year(end_date)
+    Required_years <- Required_years[which(Required_years>=2015)] #SMAP data only available after year 2015.
+    if(sum(!Required_years%in%Current_years)){
+      PEcAn.logger::logger.info("The existing SMAP.csv file doesn't contain data between start and end date!")
+      PEcAn.logger::logger.info("Please update the SMAP_gee.csv file to include the data that are missing! And then flag update_csv as TRUE to proceed!")
```

Comment:
After discovering that the soilgrids function wasn't downloading soilgrids data, I realized that I missed that the SMAP function also isn't downloading SMAP data. This seems like an even higher priority than soilgrids since SMAP data are continually updated and thus @helge22a can't use this framework until this is addressed. Could come in a future PR, but needs to come soon.

Reply:
I agree, but potentially I think building the python environment for downloading SMAP on pecan is even harder than downloading csv file directly from google earth engine (given I already created GEE code to do that), and even if someone succeed, the SMAP data after 2022 will be missing due to unknown technical issues. So I'd keep the current version unless we need to debug into the Python script on R platform.